python/tabby/tabby-data-pipeline/predict.py

```diff
@@ -0,0 +1,201 @@
+
+GPU_CONFIG = gpu.A10G()
+
+MODEL_ID = os.popen("cat /tmp/tabby_model_id").read().strip()
+LAUNCH_FLAGS = ["serve", "--model", MODEL_ID, "--port", "8000", "--device", "cuda"]
+
+
+def download_model():
+    import subprocess
+    import os
+    MODEL_ID = os.popen("cat /tmp/tabby_model_id").read().strip()
+    print(f'MODEL_ID={MODEL_ID}')
+    subprocess.run(
+        [
+            "/opt/tabby/bin/tabby",
+            "download",
+            "--model",
+            MODEL_ID,
+        ]
+    )
+
+
+image = (
+    Image.from_registry(
+        "tabbyml/tabby:0.5.5",
+        add_python="3.11",
+    )
+    .dockerfile_commands("ENTRYPOINT []")
+    .pip_install(
+        "git+https://github.com/TabbyML/tabby.git#egg=tabby-python-client&subdirectory=experimental/eval/tabby-python-client",
+        "pandas"
+    )
+    .copy_local_file(local_path="/tmp/tabby_model_id", remote_path="/tmp/tabby_model_id")
+    .run_function(download_model)
+)
+
+stub = Stub("tabby-" + MODEL_ID.split("/")[-1], image=image)
+
+
+@stub.cls(
+    gpu=GPU_CONFIG,
+    concurrency_limit=10,
+    allow_concurrent_inputs=2,
+    container_idle_timeout=60 * 10,
+    timeout=600,
+)
+class Model:
+    def __enter__(self):
+        import socket
+        import subprocess, os
+        import time
+
+        from tabby_python_client import Client
+
+        my_env = os.environ.copy()
+        my_env["TABBY_DISABLE_USAGE_COLLECTION"] = "1"
+        MODEL_ID = os.popen("cat /tmp/tabby_model_id").read().strip()
+        print(f'MODEL_ID={MODEL_ID}')
+        LAUNCH_FLAGS = ["serve", "--model", MODEL_ID, "--port", "8000", "--device", "cuda"]
+        self.launcher = subprocess.Popen(["/opt/tabby/bin/tabby"] + LAUNCH_FLAGS, env=my_env)
+        self.client = Client("http://127.0.0.1:8000", timeout=240)
+
+        # Poll until webserver at 127.0.0.1:8000 accepts connections before running inputs.
+        def webserver_ready():
+            try:
+                socket.create_connection(("127.0.0.1", 8000), timeout=1).close()
+                return True
+            except (socket.timeout, ConnectionRefusedError):
+                # Check if launcher webserving process has exited.
+                # If so, a connection can never be made.
+                retcode = self.launcher.poll()
+                if retcode is not None:
+                    raise RuntimeError(
+                        f"launcher exited unexpectedly with code {retcode}"
+                    )
+                return False
+
+        while not webserver_ready():
+            time.sleep(1.0)
+
+        print("Tabby server ready!")
+
+    def __exit__(self, _exc_type, _exc_value, _traceback):
+        self.launcher.terminate()
+
+    @method()
+    async def health(self):
+        from tabby_python_client.api.v1 import health
+
+        resp = await health.asyncio(client=self.client)
+        return resp.to_dict()
+
+    @method()
+    async def complete(self, language, crossfile_context, index, row):
```

Comment:
type annotation for every argument. For row, define a named tuple: https://docs.python.org/3/library/typing.html#typing.NamedTuple

Reply:
Do you want to list all column name of the dataframe in the namedtuple?

Reply:
For columns used for input / output

Reply:
What if there's a dynamic column? For example, "prediction" didn't exist when you first run the file and it will be added to the file.  Next time, we need to pass it through the "row"

Reply:
then it can be a optional typed field: 
https://docs.python.org/3/library/typing.html?highlight=optional#typing.Optional